components/modules/net.c

```diff
@@ -324,14 +256,103 @@ static bool post_net_accept (lnet_userdata *ud, struct tcp_pcb *newpcb) {
 }
 
 
-static err_t net_accept_cb(void *arg, struct tcp_pcb *newpcb, err_t err) {
-  lnet_userdata *ud = (lnet_userdata*)arg;
-  if (!ud || ud->type != TYPE_TCP_SERVER || !ud->pcb) return ERR_ABRT;
-  if (ud->self_ref == LUA_NOREF || ud->server.cb_accept_ref == LUA_NOREF) return ERR_ABRT;
+static void lnet_netconn_callback(struct netconn *netconn, enum netconn_evt evt, u16_t len)
+{
+  SYS_ARCH_DECL_PROTECT(lev);
+
+  if (!netconn) return;
+
+  SYS_ARCH_PROTECT(lev);
+  if (netconn->socket < 0) {
+    if (evt == NETCONN_EVT_RCVPLUS && len > 0) {
+      // data received before userdata was set up, note receive event
+      netconn->socket--;
+    }
+
+    SYS_ARCH_UNPROTECT(lev);
+    return;
+  }
+  SYS_ARCH_UNPROTECT(lev);
+
+  lnet_userdata *ud = (lnet_userdata *)netconn->socket;
+  if (!ud || !ud->netconn) return;
+
+  if (ud->type == TYPE_TCP_CLIENT || ud->type == TYPE_UDP_SOCKET) {
+    // if a previous event triggered to close the connection then skip further event processing
+    if (ud->client.closing) return;
+
+    switch (evt) {
+    case NETCONN_EVT_SENDPLUS:
+      if (ud->type == TYPE_TCP_CLIENT && ud->client.connecting) {
+        // connection established, trigger Lua callback
+        ud->client.connecting = false;
+        post_net_connected(ud);
+
+      } else if (len > 0) {
+        // data sent, trigger Lua callback
+        post_net_sent(ud);
+      }
+      break;
+
+    case NETCONN_EVT_ERROR:
+      post_net_err(ud, netconn_err(ud->netconn));
+      ud->client.closing = true;
+      break;
+
+    case NETCONN_EVT_RCVPLUS:
+      if (len > 0) {
+        // data received, collect it in Lua callback
+        post_net_recv(ud);
+      } else {
+        // signals closed connection from peer
+        post_net_err(ud, 0);
+        ud->client.closing = true;
+      }
+      break;
+
+    default:
+      break;
+    }
 
-  return post_net_accept (ud, newpcb) ? ERR_OK : ERR_ABRT;
+  } else if (ud->type == TYPE_TCP_SERVER) {
+
+    switch (evt) {
+    case NETCONN_EVT_RCVPLUS:
+      // new connection available from netconn_listen()
+      if (ud->netconn &&
+          ud->self_ref != LUA_NOREF &&
+          ud->server.cb_accept_ref != LUA_NOREF) {
```

Comment:
I'll remove the check for `self_ref` since it's not used at all in `laccept_cb`. It would open up a race condition in `net_listen()` when `netconn_listen()` was executed before the ref is actually set.

The check for `cb_accept_ref` fails when the server is in the process of being closed / shut down. Not accepting is ok in this case since the pending client request was stored in the server's netconn accept-mailbox and is going to be `tcp_abort`'ed & freed when the server netconn is finally closed and deleted.